### ISSUE TITLE
IndexOutOfBoundsException when reading XLS file

### ISSUE THREAD

**Issue by tsulej** [incanter/incanter#393]
Incanter 1.9.3/Clojure 1.9.0.

I was trying to run first example from Clojure for Data Science which failed with exception (below).

```
(xls/read-xls (str (io/resource "UK2010.xls")))
```
where excel file is taken from http://www.complex-systems.meduniwien.ac.at/elections/ElectionData/UK2010.xls

My quick investigation shows that excel is read properly. Exception is caused by `clojure.core.matrix.dataset/dataset`

Dataset contains nils at some positions. First data row looks like:

`(1.0 "Aberavon" "Wales" 2010.0 50838.0 30958.0 "" "" "" "" "" "" "" "" "" 558.0 "" "" "" "" "" 1276.0 "" "" "" "" "" "" "" "" "" "" "" 4411.0 "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" 919.0 nil nil nil nil "" "" "" "" "" "" "" "" "" 16073.0 "" 5034.0 "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" 2198.0 "" ...)`

```
java.lang.IndexOutOfBoundsException: null
 at clojure.lang.PersistentVector.arrayFor (PersistentVector.java:158)
    clojure.lang.PersistentVector.nth (PersistentVector.java:162)
    clojure.lang.RT.nth (RT.java:889)
    clojure.core.matrix.impl.persistent_vector$fn__5281.invokeStatic (persistent_vector.cljc:171)
    clojure.core.matrix.impl.persistent_vector/fn (persistent_vector.cljc:168)
    clojure.core.matrix.protocols$fn__768$G__744__775.invoke (protocols.cljc:91)
    clojure.core.matrix.impl.persistent_vector$fn__5313$iter__5314__5318$fn__5319$fn__5320$fn__5321.invoke (persistent_vector.cljc:227)
    clojure.core$mapv$fn__8088.invoke (core.clj:6832)
    clojure.lang.PersistentVector.reduce (PersistentVector.java:341)
    clojure.core$reduce.invokeStatic (core.clj:6747)
    clojure.core$mapv.invokeStatic (core.clj:6823)
    clojure.core$mapv.invoke (core.clj:6823)
    clojure.core.matrix.impl.persistent_vector$fn__5313$iter__5314__5318$fn__5319$fn__5320.invoke (persistent_vector.cljc:227)
    clojure.core.matrix.impl.persistent_vector$fn__5313$iter__5314__5318$fn__5319.invoke (persistent_vector.cljc:226)
    clojure.lang.LazySeq.sval (LazySeq.java:40)
    clojure.lang.LazySeq.seq (LazySeq.java:49)
    clojure.lang.RT.seq (RT.java:528)
    clojure.lang.LazilyPersistentVector.create (LazilyPersistentVector.java:44)
    clojure.core$vec.invokeStatic (core.clj:377)
    clojure.core$vec.invoke (core.clj:367)
    clojure.core.matrix.impl.persistent_vector$fn__5313.invokeStatic (persistent_vector.cljc:226)
    clojure.core.matrix.impl.persistent_vector/fn (persistent_vector.cljc:223)
    clojure.core.matrix.protocols$fn__1521$G__1516__1526.invoke (protocols.cljc:352)
    clojure.core.matrix.impl.sequence$fn__8163.invokeStatic (sequence.cljc:95)
    clojure.core.matrix.impl.sequence/fn (sequence.cljc:90)
    clojure.core.matrix.protocols$fn__1521$G__1516__1526.invoke (protocols.cljc:352)
    clojure.core.matrix.impl.dataset$dataset_from_rows.invokeStatic (dataset.clj:78)
    clojure.core.matrix.impl.dataset$dataset_from_rows.invoke (dataset.clj:71)
    clojure.core.matrix.dataset$dataset.invokeStatic (dataset.clj:36)
    clojure.core.matrix.dataset$dataset.invoke (dataset.clj:24)
    incanter.core$dataset.invokeStatic (core.clj:1224)
    incanter.core$dataset.invoke (core.clj:1210)
    incanter.excel$read_sheet.invokeStatic (excel.clj:93)
    incanter.excel$read_sheet.invoke (excel.clj:88)
    incanter.excel$fn__12133.invokeStatic (excel.clj:146)
    incanter.excel$fn__12133.doInvoke (excel.clj:134)
    clojure.lang.RestFn.invoke (RestFn.java:410)
    clojure.lang.MultiFn.invoke (MultiFn.java:229)
    cljds.ch1.data$eval2182$fn__2183.invoke (form-init5918460524630835521.clj:11)
    clojure.lang.MultiFn.invoke (MultiFn.java:229)
    cljds.ch1.examples$eval2344.invokeStatic (form-init5918460524630835521.clj:15)
    cljds.ch1.examples$eval2344.invoke (form-init5918460524630835521.clj:15)
    clojure.lang.Compiler.eval (Compiler.java:7062)
    clojure.lang.Compiler.eval (Compiler.java:7025)
    clojure.core$eval.invokeStatic (core.clj:3206)
    clojure.core$eval.invoke (core.clj:3202)
    clojure.main$repl$read_eval_print__8572$fn__8575.invoke (main.clj:243)
    clojure.main$repl$read_eval_print__8572.invoke (main.clj:243)
    clojure.main$repl$fn__8581.invoke (main.clj:261)
    clojure.main$repl.invokeStatic (main.clj:261)
    clojure.main$repl.doInvoke (main.clj:177)
    clojure.lang.RestFn.invoke (RestFn.java:1523)
    clojure.tools.nrepl.middleware.interruptible_eval$evaluate$fn__1536.invoke (interruptible_eval.clj:87)
    clojure.lang.AFn.applyToHelper (AFn.java:152)
    clojure.lang.AFn.applyTo (AFn.java:144)
    clojure.core$apply.invokeStatic (core.clj:657)
    clojure.core$with_bindings_STAR_.invokeStatic (core.clj:1965)
    clojure.core$with_bindings_STAR_.doInvoke (core.clj:1965)
    clojure.lang.RestFn.invoke (RestFn.java:425)
    clojure.tools.nrepl.middleware.interruptible_eval$evaluate.invokeStatic (interruptible_eval.clj:85)
    clojure.tools.nrepl.middleware.interruptible_eval$evaluate.invoke (interruptible_eval.clj:55)
    clojure.tools.nrepl.middleware.interruptible_eval$interruptible_eval$fn__1581$fn__1584.invoke (interruptible_eval.clj:222)
    clojure.tools.nrepl.middleware.interruptible_eval$run_next$fn__1576.invoke (interruptible_eval.clj:190)
    clojure.lang.AFn.run (AFn.java:22)
    java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1142)
    java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:617)
    java.lang.Thread.run (Thread.java:745)
```

**Comment by tsulej** [comment]
Worth to mention that this example works well on original versions, ie:

```
[org.clojure/clojure "1.6.0"]
[incanter/incanter "1.5.5"]
```

**Comment by joinr** [comment]
Possible dupe / related to #372  .
Note, that's unresolved.  Behavior at the moment seems to be "don't have empties/nils," which is weak IMO.  I can dig through the core.matrix stuff to try to figure out why, I think the problem is inputs going to [mp/get-cols](https://github.com/mikera/core.matrix/blob/develop/src/main/clojure/clojure/core/matrix/impl/persistent_vector.cljc#L224)  (called during dataset creation, passed "lines" read from excel zipmapped with the assumed field names from the first line) is crapping out because of the specific record not having the same number of entries as the field name, leading to a bumb call to nth, leading to index out of bounds.  That's a guess at this point.

@mikera is better positioned to speculate on the core.matrix details.

The other option is to provide better parsing and guarantees on the excel read, and ensure we're not pushing stuff that will break cmm and trip an error.

**Comment by rjungemann** [comment]
So, I re-defined the `read-sheet` method from `incanter.excel` to output some information before trying to place the rows into the dataset.

I discovered that if I `(drop-last 1 rows)` that it worked for the `UK2010.xls` spreadsheet. This is because the last row of the spreadsheet is like a summary row, and for some reason, only has one cell in the row instead of the usual 144. I think this is the root of the problem, that this row should have some `nil` cells instead of the cells being completely missing.

```clojure
(require '[incanter.excel])
(in-ns 'incanter.excel)

(defn- read-sheet [rows-it header-keywords]
  (let [colnames  (read-line-values (first rows-it))
        rows (->> (rest rows-it)
                  (map read-line-values)
                  (filter (complement empty?))
                  (drop-last 1))]
    (dataset
      (if header-keywords
        (map keyword colnames)
        colnames)
      rows)))

(incanter.excel/read-xls "UK2010.xls")
```

**Comment by tsulej** [comment]
Possibly you're right that structure of the last row causes a problem. Thanks for investigation.
This issue points out regression.

**Comment by joinr** [comment]
> So I have a feeling that the issue isn't with core.matrix, but rather with the Excel parsing. 

Great diagnosis.  This is a similar corner case I ran into when parsing the excel file for a different library.  Basically, you assume a tabular region of cells...but that's not always the case.  We'd like there to be behavior like "current region" in excel, which may select empty cells to create a rectangular region.  It could be the case that when reading cells, you get a sparse representation, which means you have to account for that during parsing (either fill in with empties or default values).  The problem is, core.matrix is assuming consistent-width rows - that is row values match the expected number of fields.  There's a call to mp/get-columns (which I think has an implementation for sequences of maps to coerce a vector of columns from the according to each field).  In the protocol definitions, there are a couple of calls to nth (none of which exist in the incanter excel wrappers), so I think that's where the indexed access error is occurring.

[core.matrix.dataset/dataset](https://github.com/mikera/core.matrix/blob/develop/src/main/clojure/clojure/core/matrix/dataset.cljc#L62) is what incanter.core/dataset wraps/delegates to, and what incanter.excel/read-sheet calls (passing the row-maps from the sheet iterator and the derived column names).  That uses the core.matrix.impl.dataset/dataset-from-row-maps to construct the dataset.  That uses map accessors and just tries to extract the known field names from the row maps (no problem, since get will return nil if there's no key, say in a sparse map like the "summary row" you mentioned).  This builds up a nested vector representation for the rows from the maps.

Then dataset-from-row-maps then passes the nested vector representation of rows to dataset-from-rows, which internally coerces the row-major representation to a column-major representation (still nested vectors) via mp/get-columns.  That's the current trace I've found.

**Comment by joinr** [comment]
@tsulej  wrote:
>Possibly you're right that structure of the last row causes a problem. Thanks for investigation.
>This issue points out regression.

Either that, or the original Excel parsing was too lenient  and possibly incorrect :)

**Comment by tsulej** [comment]
> Either that, or the original Excel parsing was too lenient and possibly incorrect :)

Yes, yes, this also should be considered here.

**Comment by joinr** [comment]
In-depth diagnosis of the problem is [here](https://gist.github.com/joinr/775aa35476f278a22183f832a318e021)

tldr: incanter is passing data to core.matrix.impl.dataset/dataset-from-rows that violates the documented assumptions core.matrix has about the data, namely that each row has an entry for each column.  In the sample data, the last row only has a single entry, so when (ultimately) nth is called on the last row for anything beyong the first column, we get said exception.

Solution would be to include validation in incanter.excel/read-sheet to validate the input prior to invoking core.matrix, or detect this condition if there's an index out of bound exception and report the error with useful information (like which row is not normalized).  Alternately, provide a way for caller to normalized data with default values, and during prep (prior to core.matrix invocation), ensure the row is normalized by adding default values if the size of a row-vector is < expected size.

Finally.....there's a behavior (complement empty?) that filters out empty rows in the xls file.  In my practice, we parse tabular, contiguous data where empty rows indicate a "Break" in the data (as with current-region in Excel), so the remainder is not processed.  Incanter will process these types of sheets and filter out "Break" rows at the moment....

**Comment by rjungemann** [comment]
If you open up the spreadsheet, the cell in the summary row is a few columns over, like this:

![image](https://user-images.githubusercontent.com/49277/43334441-65842d86-9182-11e8-8721-f6d20062fd29.png)

My instinct would be that there should be empty cells in the data imported from `read-sheet`. Essentially, `read-sheet` should indeed return the same number of cells per row, I think.

I attached the XLS file in case it is helpful in any way.

[UK2010.xls.zip](https://github.com/incanter/incanter/files/2236580/UK2010.xls.zip)

**Comment by joinr** [comment]
Yeah, that's piss poor behavior IMO.
I work with Excel a lot (most analysts use it for data entry), and my simulation has a tight data protocol that establishes contiguous tabular regions of a worksheet (default rooted at A1), with selection behavior similar to "current region" behavior in excel.  So analysts can mess with the inputs via a known medium (Excel), and then re-run the application providing the workbook as input for a project (basically a bunch of tables).

The problem the incanter.excel takes here is that it just uses each rows iterator, and doesn't check ANYTHING regarding contiguity or indexing.  [tabular-region](https://github.com/joinr/spork/blob/master/src/spork/util/excel/core.clj#L158) uses [contiguous-rows](https://github.com/joinr/spork/blob/master/src/spork/util/excel/core.clj#L86) to get the region, then fills out empty cells (i.e. cells that have no recorded value in the workbook) by [checking field count vs. row count](https://github.com/joinr/spork/blob/master/src/spork/util/excel/core.clj#L200) and normalizing the resulting row vector with nil entries for missing fields.  As I examine this, I probably need to 2x check the actual cell addresses instead of just counting the cells in each row ala row iterator.  

So, detecting where there are gaps in the data (like the singleton cell in the last example, which should be the 144th column entry) is the correct parsing behavior....

I think a solution would perform a simple comparison of row-count vs. expected field count, and if they aren't the same, unpack the sparse data from the row (using the Apache POI cell object information) to map the cell values to where they should be in the resulting row vector, then populating nil values for all other fields/columns, to get things normalized for core.matrix.

**Comment by joinr** [comment]
Hah, I forgot I do actually parse missing/non-contiguous column entries (up to the last element of the cell in the row), and [fill in blanks](https://github.com/joinr/spork/blob/master/src/spork/util/excel/core.clj#L51) up to the last cell in the row iterator sequence.  This could probably be trivially applied to incanter.excel, along with the logic that normalizes the cell entries, packaged into read-sheet.

**Comment by happy15** [comment]
any update/fix on this issue?
